### ISSUE TITLE
Can Async have Parallel?

### ISSUE THREAD

**Issue by SimY4** [typelevel/cats-effect#1150]
Hi. I'm new here. Still exploring the whole CE type class hierarchy. Feel free to tell me if the following question was already been asked before.

I'm working with `Effect[F]` which is almost at the bottom of CE typeclass hierarchy - It even has concurrent. And yet I can't seem to get myself an instance of `Parallel[F]` for it. Not until I convince type checker that my F is actually IO and I can get myself an instance of Parallel[IO].

And then I wondered if Async should probably be able to derive an instance for Parallel for any F based on its capabilities.
Parallelism is what Async give you. Applicative of async should compose asynchronously, right?

**Comment by djspiewak** [comment]
Yes it can! Or at least, `Concurrent` can (`Async` doesn't actually give you parallelism, only asynchrony). However, we didn't add one.

This was for a couple of reasons that I honestly believe are, for the most part, no longer relevant. In fact, CE3 (on the series/3.x) branch has a `Parallel` for `Concurrent`, and I'm rather happy with that decision. The interesting question is whether or not we should backport this addition to series/2.x, since it really does make a *ton* of sense.

**Comment by djspiewak** [comment]
Oh as an aside, it's important to split "asynchronous" from "parallel" and "concurrent". These are all subtly different things, and the naming of the `Parallel` typeclass doesn't help…

"Asynchronous" in this context refers to a callback-driven process. In other words, rather than *blocking* awaiting a result, you register a callback when is fired when the result is ready, releasing your carrier thread to do other work. Semantically, this is still a *sequential* action, but it is in fact asynchronous. This distinction between *semantically sequential* and "synchronous" is captured in the "fiber" terminology. We say that a series of semantically sequential actions form a *fiber*. Critically, each step of the fiber is dependent on the previous one, and so none of those steps can be reordered or executed in parallel.

"Concurrency" refers to the declarative organization of multiple fibers such that they can be mutually executed in parallel, if the underlying scheduler supports it (obviously ScalaJS does not). Confusingly, `Parallel` the typeclass is an example of such orchestration. When you do something like `(a, b).parMapN(...)`, you're using a concurrent combinator (`parMapN`) to declare that `a` and `b` are independent and should be scheduled in parallel if possible.

"Parallelism" is a function of the underlying scheduler. It literally means *these things are running simultaneously*. It's an important distinction because some platforms, such as ScalaJS or any single-threaded executor, cannot have any parallelism at all! The way I think of this is the following: concurrency is when you *declare* your program can be parallelized, while parallelism is when your concurrent program is actually evaluated partially or entirely in parallel. For the most part, you can think of parallelism as an implementation detail of Cats Effect.

Like I said, the nomenclature is pretty bad here, and the documentation even worse, but it's worth clarifying the terms since otherwise it'll be very confusing trying to discuss possibilities. :-)

**Comment by SimY4** [comment]
@djspiewak thank you for your answer. While having this question I was trying to figure out if Sync + ContextShift can actually form Parallel. I understand why Async is not really modelling parallel computations. But Parallel doesn't promise that everything will work in Parallel - it just makes sure that if it does it will compose them in a parallel manner.

Given that I was thinking of a naive version where:

parallel F ~> F is ContextShift.shift *> fa
sequential F ~> F is Sync.delay(_)

Applicative will just keep shifting things to compose later.

Haven't got anything working but seems doable. Not sure how lawful it is though.
